src/OrchardCore.Modules/OrchardCore.Setup/Filters/AutoSetupStartupFilter.cs

```diff
@@ -0,0 +1,159 @@
+                                IHttpContextAccessor accessor = scope.ServiceProvider.GetRequiredService<IHttpContextAccessor>();
+                                accessor.HttpContext.Response.Redirect("/");
+                            }
+                            else
+                            {
+                                var stringBuilder = new StringBuilder();
+                                foreach (var error in setupContext.Errors)
+                                {
+                                    stringBuilder.Append($"{error.Key} : '{error.Value}'");
+                                }
+                                _logger.LogError(T["AutoSetup failed with errors: {0}", stringBuilder.ToString()]);
+                            }
+                        }
+                    }
+                }
+            }
+
+            return next;
+        }
+
+        private SetupContext GetSetupContext(AutoSetupOptions options, ISetupService setupService, ShellSettings shellSettings)
+        {
+            var recipes = setupService.GetSetupRecipesAsync()
+                .GetAwaiter()
+                .GetResult();
+
+            var recipe = recipes.SingleOrDefault(r => r.Name == options.RecipeName);
+
+            var setupContext = new SetupContext()
+            {
+                AdminEmail = options.AdminEmail,
+                AdminPassword = options.AdminPassword,
+                AdminUsername = options.AdminUsername,
+                DatabaseConnectionString = options.DatabaseConnectionString,
+                DatabaseProvider = options.DatabaseProvider,
+                DatabaseTablePrefix = options.DatabaseTablePrefix,
+                SiteName = options.SiteName,
+                Recipe = recipe,
+                SiteTimeZone = options.SiteTimeZone,
+                Errors = new Dictionary<string, string>(),
+                ShellSettings = shellSettings
+            };
+
+            return setupContext;
+        }
+
+        private void CreateDatabase(IServiceScope scope, AutoSetupOptions options, SetupContext setupContext)
+        {
+            // maybe this should be moved to yessql?
+
+            switch (options.DatabaseProvider)
+            {
+                case "Postgres":
+                    var connectionStringBuilder = new Npgsql.NpgsqlConnectionStringBuilder(options.DatabaseConnectionString);
+                    connectionStringBuilder.Database = connectionStringBuilder["EntityAdminDatabase"].ToString();
+                    var connection = new Npgsql.NpgsqlConnection(connectionStringBuilder.ConnectionString);
+                    connection.Open();
+                    try
+                    {
+                        var cmd = connection.CreateCommand();
+                        cmd.CommandText = $"CREATE DATABASE {options.DatabaseName};";
+                        cmd.ExecuteNonQuery();
+                    }
```

Comment:
I don't remember the details but don't we need a transaction and an using block and a `transaction.Commit()`, as it is [done here in yessql](https://github.com/sebastienros/yessql/blob/master/src/YesSql.Core/Services/DbBlockIdGenerator.cs#L154).

Reply:
It's a single command, so no transaction needed.

Reply:
Okay cool